<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Data-Dictionary" data-toc-modified-id="Data-Dictionary-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Data Dictionary</a></span></li><li><span><a href="#Import-Libraries" data-toc-modified-id="Import-Libraries-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Import Libraries</a></span></li><li><span><a href="#Read-Data-and-view" data-toc-modified-id="Read-Data-and-view-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Read Data and view</a></span></li><li><span><a href="#Descriptive-stats" data-toc-modified-id="Descriptive-stats-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Descriptive stats</a></span></li><li><span><a href="#Data-Cleaning-&amp;-Preliminary-exploration-of-variables" data-toc-modified-id="Data-Cleaning-&amp;-Preliminary-exploration-of-variables-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Data Cleaning &amp; Preliminary exploration of variables</a></span><ul class="toc-item"><li><span><a href="#TIME_OCC" data-toc-modified-id="TIME_OCC-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>TIME_OCC</a></span></li><li><span><a href="#DATE_RPTD-and-DATE_OCC" data-toc-modified-id="DATE_RPTD-and-DATE_OCC-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>DATE_RPTD and DATE_OCC</a></span></li><li><span><a href="#AREA-&amp;-AREA_NAME" data-toc-modified-id="AREA-&amp;-AREA_NAME-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>AREA &amp; AREA_NAME</a></span></li><li><span><a href="#CRM_CD-&amp;-CRM_CD_DESC" data-toc-modified-id="CRM_CD-&amp;-CRM_CD_DESC-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>CRM_CD &amp; CRM_CD_DESC</a></span></li><li><span><a href="#VICT_AGE" data-toc-modified-id="VICT_AGE-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>VICT_AGE</a></span></li><li><span><a href="#VICT-SEX" data-toc-modified-id="VICT-SEX-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>VICT SEX</a></span></li><li><span><a href="#VICT_DESCENT" data-toc-modified-id="VICT_DESCENT-5.7"><span class="toc-item-num">5.7&nbsp;&nbsp;</span>VICT_DESCENT</a></span></li><li><span><a href="#PREMIS_CD-&amp;-PREMIS_DESC" data-toc-modified-id="PREMIS_CD-&amp;-PREMIS_DESC-5.8"><span class="toc-item-num">5.8&nbsp;&nbsp;</span>PREMIS_CD &amp; PREMIS_DESC</a></span></li><li><span><a href="#WEAPON_USED_CD-and-WEAPON_DESC" data-toc-modified-id="WEAPON_USED_CD-and-WEAPON_DESC-5.9"><span class="toc-item-num">5.9&nbsp;&nbsp;</span>WEAPON_USED_CD and WEAPON_DESC</a></span></li><li><span><a href="#STATUS-&amp;-STATUS_DESC" data-toc-modified-id="STATUS-&amp;-STATUS_DESC-5.10"><span class="toc-item-num">5.10&nbsp;&nbsp;</span>STATUS &amp; STATUS_DESC</a></span></li><li><span><a href="#CRM_CD-s-1,2,3-and-4" data-toc-modified-id="CRM_CD-s-1,2,3-and-4-5.11"><span class="toc-item-num">5.11&nbsp;&nbsp;</span>CRM_CD s 1,2,3 and 4</a></span></li><li><span><a href="#Location-columns:-LOCATION,-CROSS_STREET,-LAT,-LON" data-toc-modified-id="Location-columns:-LOCATION,-CROSS_STREET,-LAT,-LON-5.12"><span class="toc-item-num">5.12&nbsp;&nbsp;</span>Location columns: LOCATION, CROSS_STREET, LAT, LON</a></span></li></ul></li><li><span><a href="#Drop-columns-not-needed" data-toc-modified-id="Drop-columns-not-needed-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Drop columns not needed</a></span></li></ul></div>

# Data Dictionary

The following are the variable we are dealing with:-

- DR_NO: Division of Records Number: Official file number made up of a 2 digit year, area ID, and 5 digits. (ID var)
- API Field Name: MM/DD/YYYY.
- DATE OCC: MM/DD/YYYY.
- TIME OCC: In 24 hour military time.
- AREA: The LAPD has 21 Community Police Stations referred to as Geographic Areas within the department. These Geographic Areas are sequentially numbered from 1-21.
- AREA NAME: The 21 Geographic Areas or Patrol Divisions are also given a name designation that references a landmark or the surrounding community that it is responsible for. For example 77th Street Division is located at the intersection of South Broadway and 77th Street, serving neighborhoods in South Los Angeles.
- Rpt Dist No: A four-digit code that represents a sub-area within a Geographic Area. All crime records reference the "RD" that it occurred in for statistical comparisons.
- Crm Cd: Indicates the crime committed. (Same as Crime Code 1)
- Crm Cd Desc: Defines the Crime Code provided.
- Mocodes: Modus Operandi: Activities associated with the suspect in commission of the crime.
- Vict Age: Two character numeric.
- Vict Sex: F - Female M - Male X - Unknown.
-Vict Descent: Descent Code: A - Other Asian B - Black C - Chinese D - Cambodian F - Filipino G - Guamanian H - Hispanic/Latin/Mexican I - American Indian/Alaskan Native J - Japanese K - Korean L - Laotian O - Other P - Pacific Islander S - Samoan U - Hawaiian V - Vietnamese W - White X - Unknown Z - Asian Indian.
- Premis Cd: The type of structure, vehicle, or location where the crime took place.
- Premis Desc: Defines the Premise Code provided.
- Weapon Used Cd: The type of weapon used in the crime.
- Weapon Desc: Defines the Weapon Used Code provided.
- Status: Status of the case. (IC is the default).
- Status DEsc: Defines the Status Code provided.
- Crm Cd 1: Indicates the crime committed. Crime Code 1 is the primary and most serious one. Crime Code 2, 3, and 4 are respectively less serious offenses. Lower crime class numbers are more serious.
- Crm Cd 2: May contain a code for an additional crime, less serious than Crime Code 1.
- Crm Cd 3: May contain a code for an additional crime, less serious than Crime Code 1.
- Crm Cd 4: May contain a code for an additional crime, less serious than Crime Code 1.
- LOCATION: Street address of crime incident rounded to the nearest hundred block to maintain anonymity.
- Cross Street: Cross Street of rounded Address.
- LAT: Latitude.
- LON: Longitude.

# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

# store ml/python objects
import pickle

# Read Data and view

In [2]:
# read 2 dataframes
df1 = pd.read_csv("Crime_Data_from_2010_to_2019.csv", dtype = {"TIME OCC": str, "AREA": str})
df2 = pd.read_csv("Crime_Data_from_2020_to_Present.csv", dtype = {"TIME OCC": str, "AREA": str})

# Clean column names to concatenate

## remove spaces
df1.columns = df1.columns.str.strip().str.replace(' ', '_').str.replace("-","_")
df2.columns = df2.columns.str.strip().str.replace(' ', '_').str.replace("-","_")

## get uniform case
df1.columns = df1.columns.str.upper()
df2.columns = df2.columns.str.upper()

## concatenate and get dataframe
df = pd.concat([df1, df2])

## display
display(df.shape)

with pd.option_context("display.max_columns", None):
    display(df.head(10))


(2394173, 28)

DR_NO               DATE_RPTD                DATE_OCC TIME_OCC AREA  \
0    1307355  02/20/2010 12:00:00 AM  02/20/2010 12:00:00 AM     1350   13   
1   11401303  09/13/2010 12:00:00 AM  09/12/2010 12:00:00 AM     0045   14   
2   70309629  08/09/2010 12:00:00 AM  08/09/2010 12:00:00 AM     1515   13   
3   90631215  01/05/2010 12:00:00 AM  01/05/2010 12:00:00 AM     0150    6   
4  100100501  01/03/2010 12:00:00 AM  01/02/2010 12:00:00 AM     2100    1   
5  100100506  01/05/2010 12:00:00 AM  01/04/2010 12:00:00 AM     1650    1   
6  100100508  01/08/2010 12:00:00 AM  01/07/2010 12:00:00 AM     2005    1   
7  100100509  01/09/2010 12:00:00 AM  01/08/2010 12:00:00 AM     2100    1   
8  100100510  01/09/2010 12:00:00 AM  01/09/2010 12:00:00 AM     0230    1   
9  100100511  01/09/2010 12:00:00 AM  01/06/2010 12:00:00 AM     2100    1   

   AREA_NAME  RPT_DIST_NO  PART_1_2  CRM_CD  \
0     Newton         1385         2     900   
1    Pacific         1485         2     740   
2     Newton         1324         2     946   
3  Hollywood          646         2     900   
4    Central          176         1     122   
5    Central          162         1     442   
6    Central          182         1     330   
7    Central          157         1     230   
8    Central          171         1     230   
9    Central          132         1     341   

                                         CRM_CD_DESC         MOCODES  \
0                           VIOLATION OF COURT ORDER  0913 1814 2000   
1  VANDALISM - FELONY ($400 & OVER, ALL CHURCH VA...            0329   
2                          OTHER MISCELLANEOUS CRIME            0344   
3                           VIOLATION OF COURT ORDER  1100 0400 1402   
4                                    RAPE, ATTEMPTED            0400   
5           SHOPLIFTING - PETTY THEFT ($950 & UNDER)       0344 1402   
6                              BURGLARY FROM VEHICLE            0344   
7     ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT            0416   
8     ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT       0400 0416   
9  THEFT-GRAND ($950.01 & OVER)EXCPT,GUNS,FOWL,LI...       0344 1402   

   VICT_AGE VICT_SEX VICT_DESCENT  PREMIS_CD             PREMIS_DESC  \
0        48        M            H      501.0  SINGLE FAMILY DWELLING   
1         0        M            W      101.0                  STREET   
2         0        M            H      103.0                   ALLEY   
3        47        F            W      101.0                  STREET   
4        47        F            H      103.0                   ALLEY   
5        23        M            B      404.0        DEPARTMENT STORE   
6        46        M            H      101.0                  STREET   
7        51        M            B      710.0           OTHER PREMISE   
8        30        M            H      108.0             PARKING LOT   
9        55        M            W      710.0           OTHER PREMISE   

   WEAPON_USED_CD                                     WEAPON_DESC STATUS  \
0             NaN                                             NaN     AA   
1             NaN                                             NaN     IC   
2             NaN                                             NaN     IC   
3           102.0                                        HAND GUN     IC   
4           400.0  STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)     IC   
5             NaN                                             NaN     AA   
6             NaN                                             NaN     IC   
7           500.0                     UNKNOWN WEAPON/OTHER WEAPON     AA   
8           400.0  STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)     IC   
9             NaN                                             NaN     IC   

    STATUS_DESC  CRM_CD_1  CRM_CD_2  CRM_CD_3  CRM_CD_4  \
0  Adult Arrest     900.0       NaN       NaN       NaN   
1   Invest Cont     740.0       NaN       NaN       NaN   
2   Invest Cont     946.0     

# Descriptive stats

In [3]:
pd.set_option('float_format', '{:f}'.format)

display(df.info(null_counts=True))
display(df.describe())

<ipython-input-3-6fb9dd7ef9dc>:3: FutureWarning: null_counts is deprecated. Use show_counts instead
  display(df.info(null_counts=True))


<class 'pandas.core.frame.DataFrame'>
Int64Index: 2394173 entries, 0 to 276583
Data columns (total 28 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   DR_NO           2394173 non-null  int64  
 1   DATE_RPTD       2394173 non-null  object 
 2   DATE_OCC        2394173 non-null  object 
 3   TIME_OCC        2394173 non-null  object 
 4   AREA            2394173 non-null  object 
 5   AREA_NAME       2394173 non-null  object 
 6   RPT_DIST_NO     2394173 non-null  int64  
 7   PART_1_2        2394173 non-null  int64  
 8   CRM_CD          2394173 non-null  int64  
 9   CRM_CD_DESC     2394173 non-null  object 
 10  MOCODES         2128029 non-null  object 
 11  VICT_AGE        2394173 non-null  int64  
 12  VICT_SEX        2161085 non-null  object 
 13  VICT_DESCENT    2161034 non-null  object 
 14  PREMIS_CD       2394116 non-null  float64
 15  PREMIS_DESC     2393889 non-null  object 
 16  WEAPON_USED_CD  812466 non-null   flo

None

DR_NO    RPT_DIST_NO       PART_1_2         CRM_CD  \
count   2394173.000000 2394173.000000 2394173.000000 2394173.000000   
mean  154473833.381169    1151.772621       1.442225     507.566786   
std    32630503.736545     601.633236       0.496651     210.589220   
min         817.000000     100.000000       1.000000     110.000000   
25%   130101281.000000     642.000000       1.000000     330.000000   
50%   152115446.000000    1183.000000       1.000000     442.000000   
75%   181312160.000000    1663.000000       2.000000     626.000000   
max   910220366.000000    2199.000000       2.000000     956.000000   

            VICT_AGE      PREMIS_CD  WEAPON_USED_CD       CRM_CD_1  \
count 2394173.000000 2394116.000000   812466.000000 2394161.000000   
mean       31.550633     309.373802      370.355497     507.379157   
std        20.784620     211.617225      114.681944     210.441653   
min       -11.000000     101.000000      101.000000     110.000000   
25%        19.000000     102.000000      400.000000     330.000000   
50%        31.000000     210.000000      400.000000     442.000000   
75%        46.000000     501.000000      400.000000     626.000000   
max       120.000000     971.000000      516.000000     999.000000   

           CRM_CD_2    CRM_CD_3   CRM_CD_4            LAT            LON  
count 162688.000000 4297.000000 129.000000 2394173.000000 2394173.000000  
mean     950.416318  974.203398 977.643411      34.038289    -118.222085  
std      124.369087   80.460931  74.864962       1.162130       4.017693  
min      210.000000   93.000000 421.000000       0.000000    -118.827900  
25%      998.000000  998.000000 998.000000      34.010200    -118.435600  
50%      998.000000  998.000000 998.000000      34.061800    -118.328800  
75%      998.000000  998.000000 998.000000      34.174000    -118.277300  
max      999.000000  999.000000 999.000000      34.790700       0.000000

# Data Cleaning & Preliminary exploration of variables

In [4]:
## display
with pd.option_context("display.max_columns", None):
    display(df.head())

DR_NO               DATE_RPTD                DATE_OCC TIME_OCC AREA  \
0    1307355  02/20/2010 12:00:00 AM  02/20/2010 12:00:00 AM     1350   13   
1   11401303  09/13/2010 12:00:00 AM  09/12/2010 12:00:00 AM     0045   14   
2   70309629  08/09/2010 12:00:00 AM  08/09/2010 12:00:00 AM     1515   13   
3   90631215  01/05/2010 12:00:00 AM  01/05/2010 12:00:00 AM     0150    6   
4  100100501  01/03/2010 12:00:00 AM  01/02/2010 12:00:00 AM     2100    1   

   AREA_NAME  RPT_DIST_NO  PART_1_2  CRM_CD  \
0     Newton         1385         2     900   
1    Pacific         1485         2     740   
2     Newton         1324         2     946   
3  Hollywood          646         2     900   
4    Central          176         1     122   

                                         CRM_CD_DESC         MOCODES  \
0                           VIOLATION OF COURT ORDER  0913 1814 2000   
1  VANDALISM - FELONY ($400 & OVER, ALL CHURCH VA...            0329   
2                          OTHER MISCELLANEOUS CRIME            0344   
3                           VIOLATION OF COURT ORDER  1100 0400 1402   
4                                    RAPE, ATTEMPTED            0400   

   VICT_AGE VICT_SEX VICT_DESCENT  PREMIS_CD             PREMIS_DESC  \
0        48        M            H 501.000000  SINGLE FAMILY DWELLING   
1         0        M            W 101.000000                  STREET   
2         0        M            H 103.000000                   ALLEY   
3        47        F            W 101.000000                  STREET   
4        47        F            H 103.000000                   ALLEY   

   WEAPON_USED_CD                                     WEAPON_DESC STATUS  \
0             NaN                                             NaN     AA   
1             NaN                                             NaN     IC   
2             NaN                                             NaN     IC   
3      102.000000                                        HAND GUN     IC   
4      400.000000  STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)     IC   

    STATUS_DESC   CRM_CD_1   CRM_CD_2  CRM_CD_3  CRM_CD_4  \
0  Adult Arrest 900.000000        NaN       NaN       NaN   
1   Invest Cont 740.000000        NaN       NaN       NaN   
2   Invest Cont 946.000000        NaN       NaN       NaN   
3   Invest Cont 900.000000 998.000000       NaN       NaN   
4   Invest Cont 122.000000        NaN       NaN       NaN   

                                  LOCATION                     CROSS_STREET  \
0   300 E  GAGE                         AV                              NaN   
1          SEPULVEDA                    BL  MANCHESTER                   AV   
2  1300 E  21ST                         ST                              NaN   
3          CAHUENGA                     BL  HOLLYWOOD                    BL   
4          8TH                          ST  SAN PEDRO                    ST   

        LAT         LON  
0 33.982500 -118.269500  
1 33.959900 -118.396200  
2 34.022400 -118.252400  
3 34.101600 -118.329500  
4 34.038700 -118.248800

## TIME_OCC

In [5]:
# clean TIME_OCC

# This is imported as string to avoid importing it as a string
display(df["TIME_OCC"].head())

## convert to datetime first
df["TIME_OCC"] = pd.to_datetime(df["TIME_OCC"], format = "%H%M")

## get only hour and minute
df["TIME_OCC"] = df["TIME_OCC"].dt.time

print("after cleaning")
display(df["TIME_OCC"].head())

0    1350
1    0045
2    1515
3    0150
4    2100
Name: TIME_OCC, dtype: object

after cleaning


0    13:50:00
1    00:45:00
2    15:15:00
3    01:50:00
4    21:00:00
Name: TIME_OCC, dtype: object

## DATE_RPTD and DATE_OCC

In [6]:
# clean DATE_RPTD

##view
display(df["DATE_RPTD"].head())

## convert to datetime first
df["DATE_RPTD"] = pd.to_datetime(df["DATE_RPTD"], format = "%m/%d/%Y %H:%M:%S AM")
display(df["DATE_RPTD"].head())

## get only hour and minute
df["DATE_RPTD"] = df["DATE_RPTD"].dt.date

## convert to datetime again (after strftime it's object)
df["DATE_RPTD"] = pd.to_datetime(df["DATE_RPTD"])

print("after cleaning DATE_RPTD")
display(df["DATE_RPTD"].head())

# Clean DATE_OCC
print("For DATE_OCC")

##view
display(df["DATE_OCC"].head())

## convert to datetime first
df["DATE_OCC"] = pd.to_datetime(df["DATE_OCC"], format = "%m/%d/%Y %H:%M:%S AM")
display(df["DATE_OCC"].head())

## get only hour and minute
df["DATE_OCC"] = df["DATE_OCC"].dt.date

## convert to datetime again (after strftime it's object)
df["DATE_OCC"] = pd.to_datetime(df["DATE_OCC"])

print("after cleaning DATE_OCC")
display(df["DATE_OCC"].head())

0    02/20/2010 12:00:00 AM
1    09/13/2010 12:00:00 AM
2    08/09/2010 12:00:00 AM
3    01/05/2010 12:00:00 AM
4    01/03/2010 12:00:00 AM
Name: DATE_RPTD, dtype: object

0   2010-02-20 12:00:00
1   2010-09-13 12:00:00
2   2010-08-09 12:00:00
3   2010-01-05 12:00:00
4   2010-01-03 12:00:00
Name: DATE_RPTD, dtype: datetime64[ns]

after cleaning DATE_RPTD


0   2010-02-20
1   2010-09-13
2   2010-08-09
3   2010-01-05
4   2010-01-03
Name: DATE_RPTD, dtype: datetime64[ns]

For DATE_OCC


0    02/20/2010 12:00:00 AM
1    09/12/2010 12:00:00 AM
2    08/09/2010 12:00:00 AM
3    01/05/2010 12:00:00 AM
4    01/02/2010 12:00:00 AM
Name: DATE_OCC, dtype: object

0   2010-02-20 12:00:00
1   2010-09-12 12:00:00
2   2010-08-09 12:00:00
3   2010-01-05 12:00:00
4   2010-01-02 12:00:00
Name: DATE_OCC, dtype: datetime64[ns]

after cleaning DATE_OCC


0   2010-02-20
1   2010-09-12
2   2010-08-09
3   2010-01-05
4   2010-01-02
Name: DATE_OCC, dtype: datetime64[ns]

## AREA & AREA_NAME

In [7]:
# check data
display(df["AREA"].head())

# check value counts
display(df["AREA"].value_counts())

# change dtype
df["AREA"] = df["AREA"].astype("int")
df["AREA"] = df["AREA"].astype("category")

print("after cleaning")

# recheck value counts
display(df["AREA"].value_counts())


0    13
1    14
2    13
3     6
4     1
Name: AREA, dtype: object

12    145443
3     135559
15    114021
14    112608
18    111553
19    103687
11    100352
13    100088
9      99855
6      99111
1      98353
21     97713
17     96591
20     95417
5      92081
2      89681
10     89673
8      89145
7      88692
16     79950
4      78016
12     18346
14     16570
01     16165
03     15452
18     14813
06     14436
15     14177
20     13580
13     13451
08     13023
07     12852
02     12594
05     12344
09     12031
11     11958
19     11474
21     11106
04     10997
10     10884
17     10753
16      9578
Name: AREA, dtype: int64

after cleaning


12    163789
3     151011
14    129178
15    128198
18    126366
19    115161
1     114518
6     113547
13    113539
11    112310
9     111886
20    108997
21    108819
17    107344
5     104425
2     102275
8     102168
7     101544
10    100557
16     89528
4      89013
Name: AREA, dtype: int64

In [8]:
# Change AREA_NAME dtype
df["AREA_NAME"] = df["AREA_NAME"].astype("category")

In [9]:
# Mapping a dictionary to AREA_NAME

d_area = dict()

for (i,j) in zip(df["AREA_NAME"].values, df["AREA"].values):
    if i not in d_area.keys() and i is not np.nan:
        d_area[i] = j
        
# verify with length
display(len(d_area))
display(len(df["AREA_NAME"].value_counts()))

# display dictionary
display(d_area)

21

21

{'Newton': 13,
 'Pacific': 14,
 'Hollywood': 6,
 'Central': 1,
 'Northeast': 11,
 'Hollenbeck': 4,
 'Southwest': 3,
 'Southeast': 18,
 'Rampart': 2,
 'Olympic': 20,
 'Wilshire': 7,
 '77th Street': 12,
 'Harbor': 5,
 'West LA': 8,
 'Van Nuys': 9,
 'West Valley': 10,
 'Mission': 19,
 'Topanga': 21,
 'N Hollywood': 15,
 'Foothill': 16,
 'Devonshire': 17}

## CRM_CD & CRM_CD_DESC

In [10]:
# convert to category
df["CRM_CD"] = df["CRM_CD"].astype("category")

# check value_counts
display(df["CRM_CD"].value_counts())
        
df["CRM_CD_DESC"] = df["CRM_CD_DESC"].astype("category")

# check value_counts
display(df["CRM_CD_DESC"].value_counts())

624    213068
510    189938
330    180102
310    165285
440    164982
        ...  
432         3
926         2
453         1
905         1
904         1
Name: CRM_CD, Length: 143, dtype: int64

BATTERY - SIMPLE ASSAULT                                   213068
VEHICLE - STOLEN                                           189938
BURGLARY FROM VEHICLE                                      180102
BURGLARY                                                   165285
THEFT PLAIN - PETTY ($950 & UNDER)                         164982
                                                            ...  
BLOCKING DOOR INDUCTION CENTER                                  3
TRAIN WRECKING                                                  2
FIREARMS EMERGENCY PROTECTIVE ORDER (FIREARMS EPO)              1
FIREARMS TEMPORARY RESTRAINING ORDER (TEMP FIREARMS RO)         1
DRUNK ROLL - ATTEMPT                                            1
Name: CRM_CD_DESC, Length: 143, dtype: int64

In [11]:
# Mapping a dictionary

d_crm = dict()

for (i,j) in zip(df["CRM_CD_DESC"].values, df["CRM_CD"].values):
    if i not in d_crm.keys() and i is not np.nan:
        d_crm[i] = j
        
# verify with length
display(len(d_crm))
display(len(df["CRM_CD_DESC"].value_counts()))

# display dictionary
display(d_crm)

143

143

{'VIOLATION OF COURT ORDER': 900,
 'VANDALISM - FELONY ($400 & OVER, ALL CHURCH VANDALISMS)': 740,
 'OTHER MISCELLANEOUS CRIME': 946,
 'RAPE, ATTEMPTED': 122,
 'SHOPLIFTING - PETTY THEFT ($950 & UNDER)': 442,
 'BURGLARY FROM VEHICLE': 330,
 'ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT': 230,
 'THEFT-GRAND ($950.01 & OVER)EXCPT,GUNS,FOWL,LIVESTK,PROD': 341,
 'BATTERY - SIMPLE ASSAULT': 624,
 'ROBBERY': 210,
 'BOMB SCARE': 755,
 'CHILD NEGLECT (SEE 300 W.I.C.)': 237,
 'INTIMATE PARTNER - AGGRAVATED ASSAULT': 236,
 'INTIMATE PARTNER - SIMPLE ASSAULT': 626,
 'THEFT PLAIN - PETTY ($950 & UNDER)': 440,
 'CRIMINAL THREATS - NO WEAPON DISPLAYED': 930,
 'ATTEMPTED ROBBERY': 220,
 'VANDALISM - MISDEAMEANOR ($399 OR UNDER)': 745,
 'BURGLARY': 310,
 'ARSON': 648,
 'RAPE, FORCIBLE': 121,
 'BRANDISH WEAPON': 761,
 'THROWING OBJECT AT MOVING VEHICLE': 647,
 'SHOPLIFTING-GRAND THEFT ($950.01 & OVER)': 343,
 'CHILD ABUSE (PHYSICAL) - SIMPLE ASSAULT': 627,
 'SHOTS FIRED AT INHABITED DWELLING': 251,
 

**Notes**:-
    
- Here we can create one more column: severity crimes committed
- the severity will be found by summing up the codes.
- To control for scales, we can divide the code by 100 and convert to int (so anything 0-100 will be 1, 100-199 will be 2, etc).
- Also, note that lower codes mean severe crimes, so we subtract this number from the highest code obtainable (say 900-999 is the final code in our dataset, then we do 11-x where x is the current code)
- This approach can also be used for our other CRIME_CD data column, instead of using descriptions

In [12]:
## bucket
df["CRM_CD"] = pd.cut(df["CRM_CD"], [0, 100, 200, 300 ,400 ,500 ,600 ,700, 800, 900, 1000], labels = [1,2,3,4,5,6,7,8,9,10], ordered = True)

## change dtype
df["CRM_CD"] = df["CRM_CD"].astype("int")

## subtract for severity
df["CRM_CD"] = 11 - df["CRM_CD"]

## view
display(df["CRM_CD"].head())
display(df["CRM_CD"].value_counts())

0    2
1    3
2    1
3    2
4    9
Name: CRM_CD, dtype: int32

7    643410
4    426526
6    351008
3    257911
8    249027
5    194285
1    156285
2     99627
9     16094
Name: CRM_CD, dtype: int64

## VICT_AGE

In [13]:
# There are some negative values in age which we clean
df["VICT_AGE"].describe()

count   2394173.000000
mean         31.550633
std          20.784620
min         -11.000000
25%          19.000000
50%          31.000000
75%          46.000000
max         120.000000
Name: VICT_AGE, dtype: float64

In [14]:
# Check number of negative values
print((df["VICT_AGE"] < 0).mean())

# Replace negative values with 0
df.loc[df["VICT_AGE"] < 0, "VICT_AGE"] = 0

# review
display(df["VICT_AGE"].describe())

0.00026856872916034057


count   2394173.000000
mean         31.551282
std          20.783574
min           0.000000
25%          19.000000
50%          31.000000
75%          46.000000
max         120.000000
Name: VICT_AGE, dtype: float64

## VICT SEX

In [15]:
# check head
display(df["VICT_SEX"].head())

# check value counts
display(df["VICT_SEX"].value_counts())

# this column has missing vals. Club into X
df["VICT_SEX"].fillna("X", inplace = True)

# couple X, H, N, - together into X (Unknown)

df["VICT_SEX"].replace({"H": "X",
                       "N": "X",
                       "-": "X"}, 
                      inplace = True)

# recheck value counts
print("after changing")
display(df["VICT_SEX"].value_counts())


# convert to category
df["VICT_SEX"] = df["VICT_SEX"].astype("category")

0    M
1    M
2    M
3    F
4    F
Name: VICT_SEX, dtype: object

M    1093005
F     989833
X      78131
H         98
N         17
-          1
Name: VICT_SEX, dtype: int64

after changing


M    1093005
F     989833
X     311335
Name: VICT_SEX, dtype: int64

## VICT_DESCENT

In [16]:
display(df["VICT_DESCENT"].head())

# this column has missing values. Club missing into X (unknown) category
df["VICT_DESCENT"].fillna("X", inplace = True)

# check value counts
display(df["VICT_DESCENT"].value_counts())


#check nunique
display(df["VICT_DESCENT"].nunique())

# we can club O (other) and - into X too
df["VICT_DESCENT"].replace({"O":"X",
                            "-":"X"},
                          inplace = True)

print("after cleaning")

# recheck value_counts
display(df["VICT_DESCENT"].value_counts())

# convert to category
df["VICT_DESCENT"] = df["VICT_DESCENT"].astype("category")

0    H
1    W
2    H
3    W
4    H
Name: VICT_DESCENT, dtype: object

H    811317
W    569776
B    375047
X    336883
O    225305
A     57124
K     10405
F      3376
C      1708
I      1113
J       682
V       400
P       395
U       226
Z       201
G       107
S        45
D        34
L        26
-         3
Name: VICT_DESCENT, dtype: int64

20

after cleaning


H    811317
W    569776
X    562191
B    375047
A     57124
K     10405
F      3376
C      1708
I      1113
J       682
V       400
P       395
U       226
Z       201
G       107
S        45
D        34
L        26
Name: VICT_DESCENT, dtype: int64

In [17]:
# Convert to name
races = {'A' : 'Other Asian', 'B' : 'Black', 'C' : 'Chinese', 'D' : 'Cambodian', 'F' : 'Filipino', 'G' : 'Guamanian', 'H' : 'Hispanic/Latin/Mexican', \
         'I' : 'American Indian/Alaskan Native', 'J' : 'Japanese', 'K' : 'Korean', 'L' : 'Laotian', 'O' : 'Other', 'P' : 'Pacific Islander', 'S' : 'Samoan', \
         'U' : 'Hawaiian', 'V' : 'Vietnamese', 'W' : 'White', 'X' : 'Unknown', 'Z' : 'Asian Indian'}
df['VICTIM_RACE'] = df['VICT_DESCENT'].map(races)
display(df["VICTIM_RACE"].head())

0    Hispanic/Latin/Mexican
1                     White
2    Hispanic/Latin/Mexican
3                     White
4    Hispanic/Latin/Mexican
Name: VICTIM_RACE, dtype: category
Categories (18, object): ['Other Asian', 'Black', 'Chinese', 'Cambodian', ..., 'Vietnamese', 'White', 'Unknown', 'Asian Indian']

## PREMIS_CD & PREMIS_DESC

In [18]:
# PREMIS_CD
display(df["PREMIS_CD"].head())

## check value counts
with pd.option_context("display.max_rows", None):
    display(df["PREMIS_CD"].value_counts())

## check nulls
display(df["PREMIS_CD"].isnull().sum())

## convert dtypes: first int then to category
df["PREMIS_CD"] = df["PREMIS_CD"].astype("Int64")
df["PREMIS_CD"] = df["PREMIS_CD"].astype("category")

# PREMIS_DESC
display(df["PREMIS_DESC"].head())

## check value counts
with pd.option_context("display.max_rows", None):
    display(df["PREMIS_DESC"].value_counts())

## check nulls
display(df["PREMIS_DESC"].isnull().sum())

## convert dtypes
df["PREMIS_DESC"] = df["PREMIS_DESC"].astype("category") 




0   501.000000
1   101.000000
2   103.000000
3   101.000000
4   103.000000
Name: PREMIS_CD, dtype: float64

101.000000    544366
501.000000    475251
502.000000    299265
108.000000    170457
102.000000    118772
203.000000    109381
122.000000     89249
104.000000     48084
707.000000     43684
210.000000     32435
404.000000     30635
402.000000     24095
123.000000     19041
406.000000     17798
109.000000     16614
121.000000     16303
103.000000     16039
710.000000     15327
721.000000     13475
301.000000     12200
403.000000     11969
503.000000     11461
504.000000     10453
720.000000      8729
401.000000      8644
212.000000      8266
704.000000      8055
221.000000      7608
702.000000      7467
602.000000      7162
505.000000      6127
726.000000      6098
405.000000      5902
202.000000      5873
701.000000      5858
717.000000      5650
119.000000      5304
507.000000      5201
124.000000      5110
708.000000      5099
735.000000      4799
248.000000      4301
222.000000      4252
722.000000      3998
118.000000      3932
725.000000      3714
727.000000      3651
116.000000   

57

0    SINGLE FAMILY DWELLING
1                    STREET
2                     ALLEY
3                    STREET
4                     ALLEY
Name: PREMIS_DESC, dtype: object

STREET                                                             544366
SINGLE FAMILY DWELLING                                             475251
MULTI-UNIT DWELLING (APARTMENT, DUPLEX, ETC)                       299265
PARKING LOT                                                        170457
SIDEWALK                                                           118772
OTHER BUSINESS                                                     109381
VEHICLE, PASSENGER/TRUCK                                            89249
DRIVEWAY                                                            48084
GARAGE/CARPORT                                                      43684
RESTAURANT/FAST FOOD                                                32435
DEPARTMENT STORE                                                    30635
MARKET                                                              24095
PARKING UNDERGROUND/BUILDING                                        19041
OTHER STORE                           

284

**Note**

- here there are more nulls in PREMIS_DESC and also unique values are lesser in PREMIS_DESC

In [19]:
# Mapping a dictionary

d_premis = dict()

for (i,j) in zip(df["PREMIS_DESC"].values, df["PREMIS_CD"].values):
    if i not in d_premis.keys() and i is not np.nan:
        d_premis[i] = j
        
# verify with length
display(len(d_premis))
display(len(df["PREMIS_DESC"].value_counts()))

# display dictionary
display(d_premis)

320

320

{'SINGLE FAMILY DWELLING': 501,
 'STREET': 101,
 'ALLEY': 103,
 'DEPARTMENT STORE': 404,
 'OTHER PREMISE': 710,
 'PARKING LOT': 108,
 'SIDEWALK': 102,
 'LIBRARY': 738,
 'MULTI-UNIT DWELLING (APARTMENT, DUPLEX, ETC)': 502,
 'RESTAURANT/FAST FOOD': 210,
 'BAR/COCKTAIL/NIGHTCLUB': 207,
 'HOTEL': 503,
 'OTHER BUSINESS': 203,
 'OTHER STORE': 406,
 'GOVERNMENT FACILITY (FEDERAL,STATE, COUNTY & CITY)': 725,
 'VEHICLE, PASSENGER/TRUCK': 122,
 'MUNICIPAL BUS LINE INCLUDES LADOT/DASH': 802,
 'PAWN SHOP': 211,
 'OTHER RESIDENCE': 504,
 'JEWELRY STORE': 201,
 'WAREHOUSE': 213,
 'SPECIALTY SCHOOL/OTHER': 729,
 'CHURCH/CHAPEL (CHANGED 03-03 FROM CHURCH/TEMPLE)': 708,
 'BUS STOP': 124,
 'OFFICE BUILDING/OFFICE': 702,
 'CONSTRUCTION SITE': 118,
 'PARKING UNDERGROUND/BUILDING': 123,
 'HOSPITAL': 701,
 'DRIVEWAY': 104,
 'NIGHT CLUB (OPEN EVENINGS ONLY)': 735,
 'DRUG STORE': 403,
 'LIQUOR STORE': 202,
 'BANK': 602,
 'MARKET': 402,
 'TRAIN DEPOT/TERMINAL, OTHER THAN MTA': 215,
 'PUBLIC RESTROOM/OUTSIDE*':

In [20]:
# check the codes that are not mapped
set(df["PREMIS_CD"]) - set(d_premis.values())

{nan, 226, 256, 418, 805, 838}

**Notes**

- The codes can be included if we group the data.
- Also, there is an OTHER PREMISE category, where the nulls can be clubbed and the current column used

In [21]:
# Clean

## filter where there are above codes
l = list(set(df["PREMIS_CD"]) - set(d_premis.values()))
display(l)


## update descriptions for the above codes
df.loc[df["PREMIS_CD"].isin(l), "PREMIS_DESC"] = "OTHER PREMISE"

## view to check
display(df["PREMIS_DESC"].isnull().sum())
display(df.loc[df["PREMIS_CD"].isin(l), "PREMIS_DESC"])

[nan, 256, 226, 418, 805, 838]

0

6590      OTHER PREMISE
32148     OTHER PREMISE
67361     OTHER PREMISE
68300     OTHER PREMISE
71555     OTHER PREMISE
              ...      
267160    OTHER PREMISE
267814    OTHER PREMISE
268521    OTHER PREMISE
269988    OTHER PREMISE
272035    OTHER PREMISE
Name: PREMIS_DESC, Length: 301, dtype: category
Categories (320, object): ['7TH AND METRO CENTER (NOT LINE SPECIFIC)', 'ABANDONED BUILDING ABANDONED HOUSE', 'ABATEMENT LOCATION', 'ABORTION CLINIC/ABORTION FACILITY*', ..., 'WAREHOUSE', 'WATER FACILITY', 'WEBSITE', 'YARD (RESIDENTIAL/BUSINESS)']

## WEAPON_USED_CD and WEAPON_DESC

In [22]:
# WEAPON_USED_CD
display(df["WEAPON_USED_CD"].head())

## check value counts
with pd.option_context("display.max_rows", None):
    display(df["WEAPON_USED_CD"].value_counts())

## check nulls
display(df["WEAPON_USED_CD"].isnull().sum())

## convert dtypes: first int then to category
df["WEAPON_USED_CD"] = df["WEAPON_USED_CD"].astype("Int64")
df["WEAPON_USED_CD"] = df["WEAPON_USED_CD"].astype("category")

# WEAPON_DESC
display(df["WEAPON_DESC"].head())

## check value counts
with pd.option_context("display.max_rows", None):
    display(df["WEAPON_DESC"].value_counts())

## check nulls
display(df["WEAPON_DESC"].isnull().sum())

## convert dtypes
df["WEAPON_DESC"] = df["WEAPON_DESC"].astype("category") 

## display length
display(len(df["WEAPON_USED_CD"].value_counts()))
display(len(df["WEAPON_DESC"].value_counts()))


0          NaN
1          NaN
2          NaN
3   102.000000
4   400.000000
Name: WEAPON_USED_CD, dtype: float64

400.000000    485450
500.000000     68333
511.000000     65935
102.000000     40017
109.000000     15099
200.000000     15000
207.000000     11538
106.000000      9984
307.000000      8274
101.000000      6039
212.000000      5879
306.000000      5528
308.000000      5291
304.000000      5103
512.000000      5034
204.000000      4983
302.000000      4761
205.000000      4435
201.000000      3956
113.000000      3934
114.000000      3256
312.000000      3067
301.000000      2833
218.000000      2546
515.000000      2049
311.000000      1925
219.000000      1902
223.000000      1761
215.000000      1523
216.000000      1353
104.000000      1326
310.000000      1182
107.000000      1134
103.000000      1028
305.000000       984
513.000000       772
506.000000       753
309.000000       701
206.000000       635
501.000000       580
221.000000       553
503.000000       542
303.000000       484
514.000000       434
211.000000       431
112.000000       368
510.000000       347
504.000000   

1581707

0                                               NaN
1                                               NaN
2                                               NaN
3                                          HAND GUN
4    STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)
Name: WEAPON_DESC, dtype: object

STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)    485450
UNKNOWN WEAPON/OTHER WEAPON                        68333
VERBAL THREAT                                      65935
HAND GUN                                           40017
SEMI-AUTOMATIC PISTOL                              15099
KNIFE WITH BLADE 6INCHES OR LESS                   15000
OTHER KNIFE                                        11538
UNKNOWN FIREARM                                     9984
VEHICLE                                             8274
REVOLVER                                            6039
BOTTLE                                              5879
ROCK/THROWN OBJECT                                  5528
STICK                                               5291
CLUB/BAT                                            5103
MACE/PEPPER SPRAY                                   5034
FOLDING KNIFE                                       4983
BLUNT INSTRUMENT                                    4761
KITCHEN KNIFE                  

1581708

80

79

**Note**

- There are more nulls in description than in code

In [23]:
# Mapping a dictionary

d_weapon = dict()

for (i,j) in zip(df["WEAPON_DESC"].values, df["WEAPON_USED_CD"].values):
    if i not in d_weapon.keys() and i is not np.nan:
        d_weapon[i] = j
        
# verify with length
display(len(d_weapon))
display(len(df["WEAPON_DESC"].value_counts()))

# display dictionary
display(d_weapon)

79

79

{'HAND GUN': 102,
 'STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)': 400,
 'UNKNOWN WEAPON/OTHER WEAPON': 500,
 'VERBAL THREAT': 511,
 'SIMULATED GUN': 113,
 'FOLDING KNIFE': 204,
 'OTHER KNIFE': 207,
 'MACE/PEPPER SPRAY': 512,
 'STICK': 308,
 'UNKNOWN FIREARM': 106,
 'SWITCH BLADE': 206,
 'KNIFE WITH BLADE 6INCHES OR LESS': 200,
 'BOARD': 309,
 'BLUNT INSTRUMENT': 302,
 'SEMI-AUTOMATIC PISTOL': 109,
 'VEHICLE': 307,
 'OTHER CUTTING INSTRUMENT': 218,
 'SCISSORS': 216,
 'BOTTLE': 212,
 'KNIFE WITH BLADE OVER 6 INCHES IN LENGTH': 201,
 'BELT FLAILING INSTRUMENT/CHAIN': 301,
 'AUTOMATIC WEAPON/SUB-MACHINE GUN': 108,
 'CLUB/BAT': 304,
 'BRASS KNUCKLES': 303,
 'KITCHEN KNIFE': 205,
 'AIR PISTOL/REVOLVER/RIFLE/BB GUN': 114,
 'SCREWDRIVER': 219,
 'FIRE': 506,
 'STUN GUN': 513,
 'RIFLE': 103,
 'RAZOR BLADE': 210,
 'DIRK/DAGGER': 203,
 'CAUSTIC CHEMICAL/POISON': 503,
 'SHOTGUN': 104,
 'ICE PICK': 214,
 'BOMB THREAT': 501,
 'ROCK/THROWN OBJECT': 306,
 'REVOLVER': 101,
 'OTHER FIREARM': 107,
 'DE

In [24]:
# check the codes that are not mapped
set(df["WEAPON_USED_CD"]) - set(d_weapon.values())

{nan, 222}

**Note**

- Like the case for PREMIS, we can try to club the 222 with the rest: essentially, unknwon type cutting instrument
- Also, we can have UNKNOWN WEAPON/OTHER WEAPON category. All nulls will be clubbed there and the column desc more readily used

In [25]:
# Clean

## filter where there are above codes
l = list(set(df["WEAPON_USED_CD"]) - set(d_weapon.values()))
display(l)


## update descriptions for the above codes
df.loc[df["WEAPON_USED_CD"].isin(l), "WEAPON_DESC"] = "UNKNOWN WEAPON/OTHER WEAPON"

## view to check
display(df["WEAPON_DESC"].isnull().sum())
display(df.loc[df["WEAPON_USED_CD"].isin(l), "WEAPON_DESC"])

[nan, 222]

0

0         UNKNOWN WEAPON/OTHER WEAPON
1         UNKNOWN WEAPON/OTHER WEAPON
2         UNKNOWN WEAPON/OTHER WEAPON
5         UNKNOWN WEAPON/OTHER WEAPON
6         UNKNOWN WEAPON/OTHER WEAPON
                     ...             
276578    UNKNOWN WEAPON/OTHER WEAPON
276579    UNKNOWN WEAPON/OTHER WEAPON
276580    UNKNOWN WEAPON/OTHER WEAPON
276581    UNKNOWN WEAPON/OTHER WEAPON
276583    UNKNOWN WEAPON/OTHER WEAPON
Name: WEAPON_DESC, Length: 1581708, dtype: category
Categories (79, object): ['AIR PISTOL/REVOLVER/RIFLE/BB GUN', 'ANTIQUE FIREARM', 'ASSAULT WEAPON/UZI/AK47/ETC', 'AUTOMATIC WEAPON/SUB-MACHINE GUN', ..., 'UNKNOWN WEAPON/OTHER WEAPON', 'UZI SEMIAUTOMATIC ASSAULT RIFLE', 'VEHICLE', 'VERBAL THREAT']

## STATUS & STATUS_DESC

In [26]:
# STATUS
display(df["STATUS"].head())

## check value counts
with pd.option_context("display.max_rows", None):
    display(df["STATUS"].value_counts())

## check nulls
display(df["STATUS"].isnull().sum())

## convert dtypes
df["STATUS"] = df["STATUS"].astype("category")

# STATUS_DESC
display(df["STATUS_DESC"].head())

## check value counts
with pd.option_context("display.max_rows", None):
    display(df["STATUS_DESC"].value_counts())

## check nulls
display(df["STATUS_DESC"].isnull().sum())

## convert dtypes
df["STATUS_DESC"] = df["STATUS_DESC"].astype("category") 

## display length
display(len(df["STATUS"].value_counts()))
display(len(df["STATUS_DESC"].value_counts()))

0    AA
1    IC
2    IC
3    IC
4    IC
Name: STATUS, dtype: object

IC    1843723
AO     282732
AA     245137
JA      16818
JO       5728
CC         29
19          1
13          1
TH          1
Name: STATUS, dtype: int64

3

0    Adult Arrest
1     Invest Cont
2     Invest Cont
3     Invest Cont
4     Invest Cont
Name: STATUS_DESC, dtype: object

Invest Cont     1843723
Adult Other      282732
Adult Arrest     245137
Juv Arrest        16818
Juv Other          5728
UNK                  35
Name: STATUS_DESC, dtype: int64

0

9

6

**Notes**

- Looking at the above counts, we can directly map the description to codes as initials. Also, UNK category is used for the codes CC, TH, 19, 13 and the null values. Thus we can drop STATUS and use STATUS_DESC for our purpose

## CRM_CD s 1,2,3 and 4

In [27]:
# check nulls
display(df["CRM_CD_1"].isnull().sum())
display(df["CRM_CD_2"].isnull().sum())
display(df["CRM_CD_3"].isnull().sum())
display(df["CRM_CD_4"].isnull().sum())

12

2231485

2389876

2394044

**Notes:-**
- Same method is used for cleaning as of CRM_CD
- Also, all codes are added for creating new column, SEVERITY, represnting cumulative crime

In [28]:
# update values for CRM_CD_2, 3 and 4

# CRM_CD_2
##fillna
df["CRM_CD_2"].fillna(1001, inplace = True)

## bucket
df["CRM_CD_2"] = pd.cut(df["CRM_CD_2"], [0, 100, 200, 300 ,400 ,500 ,600 ,700, 800, 900, 1000, 1100],
                        labels = [1,2,3,4,5,6,7,8,9, 10, 11], ordered = True)

## change dtype
df["CRM_CD_2"] = df["CRM_CD_2"].astype("int")

## subtract for severity
df["CRM_CD_2"] = 11 - df["CRM_CD_2"]

## view
df["CRM_CD_2"].head()

0    0
1    0
2    0
3    1
4    0
Name: CRM_CD_2, dtype: int32

In [29]:
# CRM_CD_3
##fillna
df["CRM_CD_3"].fillna(1001, inplace = True)

## bucket
df["CRM_CD_3"] = pd.cut(df["CRM_CD_3"], [0, 100, 200, 300 ,400 ,500 ,600 ,700, 800, 900, 1000, 1100], 
                        labels = [1,2,3,4,5,6,7,8,9,10, 11], ordered = True)

## change dtype
df["CRM_CD_3"] = df["CRM_CD_3"].astype("int")

## subtract for severity
df["CRM_CD_3"] = 11 - df["CRM_CD_3"]

## view
df["CRM_CD_3"].head()

0    0
1    0
2    0
3    0
4    0
Name: CRM_CD_3, dtype: int32

In [30]:
# CRM_CD_4
##fillna
df["CRM_CD_4"].fillna(1001, inplace = True)

## bucket
df["CRM_CD_4"] = pd.cut(df["CRM_CD_4"], [0, 100, 200, 300 ,400 ,500 ,600 ,700, 800, 900, 1000, 1100], 
                        labels = [1,2,3,4,5,6,7,8,9,10,11], ordered = True)

## change dtype
df["CRM_CD_4"] = df["CRM_CD_4"].astype("int")

## subtract for severity
df["CRM_CD_4"] = 11 - df["CRM_CD_4"]

## view
df["CRM_CD_4"].head()

0    0
1    0
2    0
3    0
4    0
Name: CRM_CD_4, dtype: int32

In [31]:
# Add the columns creating new column severity
df["SEVERITY"] = df["CRM_CD"] + df["CRM_CD_2"] + df["CRM_CD_3"] + df["CRM_CD_4"]
df["SEVERITY"].head()

0    2
1    3
2    1
3    3
4    9
Name: SEVERITY, dtype: int32

In [32]:
# can drop CRM_CD_2 to 4

## Location columns: LOCATION, CROSS_STREET, LAT, LON

In [33]:
# LOCATION:
display(df["LOCATION"].head())

# CROSS_STREET
display(df["CROSS_STREET"].head())

# LAT and LON
display(df["LAT"].head())
display(df["LON"].head())

0     300 E  GAGE                         AV
1            SEPULVEDA                    BL
2    1300 E  21ST                         ST
3            CAHUENGA                     BL
4            8TH                          ST
Name: LOCATION, dtype: object

0                                NaN
1    MANCHESTER                   AV
2                                NaN
3    HOLLYWOOD                    BL
4    SAN PEDRO                    ST
Name: CROSS_STREET, dtype: object

0   33.982500
1   33.959900
2   34.022400
3   34.101600
4   34.038700
Name: LAT, dtype: float64

0   -118.269500
1   -118.396200
2   -118.252400
3   -118.329500
4   -118.248800
Name: LON, dtype: float64

**Notes**:-
    
- All the above columns except location can be deleted and replaced by zipcodes

In [34]:
# change dtype of location
df["LOCATION"] = df["LOCATION"].astype("category")

# Drop columns not needed

In [35]:
# view final columns and data

display(df.columns)

with pd.option_context("display.max_columns", None):
    display(df.head())

Index(['DR_NO', 'DATE_RPTD', 'DATE_OCC', 'TIME_OCC', 'AREA', 'AREA_NAME',
       'RPT_DIST_NO', 'PART_1_2', 'CRM_CD', 'CRM_CD_DESC', 'MOCODES',
       'VICT_AGE', 'VICT_SEX', 'VICT_DESCENT', 'PREMIS_CD', 'PREMIS_DESC',
       'WEAPON_USED_CD', 'WEAPON_DESC', 'STATUS', 'STATUS_DESC', 'CRM_CD_1',
       'CRM_CD_2', 'CRM_CD_3', 'CRM_CD_4', 'LOCATION', 'CROSS_STREET', 'LAT',
       'LON', 'VICTIM_RACE', 'SEVERITY'],
      dtype='object')

DR_NO  DATE_RPTD   DATE_OCC  TIME_OCC AREA  AREA_NAME  RPT_DIST_NO  \
0    1307355 2010-02-20 2010-02-20  13:50:00   13     Newton         1385   
1   11401303 2010-09-13 2010-09-12  00:45:00   14    Pacific         1485   
2   70309629 2010-08-09 2010-08-09  15:15:00   13     Newton         1324   
3   90631215 2010-01-05 2010-01-05  01:50:00    6  Hollywood          646   
4  100100501 2010-01-03 2010-01-02  21:00:00    1    Central          176   

   PART_1_2  CRM_CD                                        CRM_CD_DESC  \
0         2       2                           VIOLATION OF COURT ORDER   
1         2       3  VANDALISM - FELONY ($400 & OVER, ALL CHURCH VA...   
2         2       1                          OTHER MISCELLANEOUS CRIME   
3         2       2                           VIOLATION OF COURT ORDER   
4         1       9                                    RAPE, ATTEMPTED   

          MOCODES  VICT_AGE VICT_SEX VICT_DESCENT PREMIS_CD  \
0  0913 1814 2000        48        M            H       501   
1            0329         0        M            W       101   
2            0344         0        M            H       103   
3  1100 0400 1402        47        F            W       101   
4            0400        47        F            H       103   

              PREMIS_DESC WEAPON_USED_CD  \
0  SINGLE FAMILY DWELLING            NaN   
1                  STREET            NaN   
2                   ALLEY            NaN   
3                  STREET            102   
4                   ALLEY            400   

                                      WEAPON_DESC STATUS   STATUS_DESC  \
0                     UNKNOWN WEAPON/OTHER WEAPON     AA  Adult Arrest   
1                     UNKNOWN WEAPON/OTHER WEAPON     IC   Invest Cont   
2                     UNKNOWN WEAPON/OTHER WEAPON     IC   Invest Cont   
3                                        HAND GUN     IC   Invest Cont   
4  STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)     IC   Invest Cont   

    CRM_CD_1  CRM_CD_2  CRM_CD_3  CRM_CD_4  \
0 900.000000         0         0         0   
1 740.000000         0         0         0   
2 946.000000         0         0         0   
3 900.000000         1         0         0   
4 122.000000         0         0         0   

                                  LOCATION                     CROSS_STREET  \
0   300 E  GAGE                         AV                              NaN   
1          SEPULVEDA                    BL  MANCHESTER                   AV   
2  1300 E  21ST                         ST                              NaN   
3          CAHUENGA                     BL  HOLLYWOOD                    BL   
4          8TH                          ST  SAN PEDRO                    ST   

        LAT         LON             VICTIM_RACE  SEVERITY  
0 33.982500 -118.269500  Hispanic/Latin/Mexican         2  
1 33.959900 -118.396200                   White         3  
2 34.022400 -118.252400  Hispanic/Latin/Mexican         1  
3 34.101600 -118.329500                   White         3  
4 34.038700 -118.248800  Hispanic/Latin/Mexican         9

In [36]:
# drop columns: suggested is to drop the following
df.drop(["RPT_DIST_NO",
        "PART_1_2", "CRM_CD_DESC",
         "AREA",
         "CRM_CD_1", "CRM_CD_2", "CRM_CD_3", "CRM_CD_4",
        "MOCODES",
        "PREMIS_CD",
        "WEAPON_USED_CD",
        "STATUS",
         "VICT_DESCENT",
        "CROSS_STREET"], axis = 1, inplace = True)

In [37]:
# Rename CRM_CD to CRM_1_SEVERITY
df.rename(columns = {"CRM_CD": "CRM_1_SEVERITY"}, inplace = True)

In [38]:
# view final dataset
display(df.columns)
with pd.option_context("display.max_columns", None):
    display(df.head())

Index(['DR_NO', 'DATE_RPTD', 'DATE_OCC', 'TIME_OCC', 'AREA_NAME',
       'CRM_1_SEVERITY', 'VICT_AGE', 'VICT_SEX', 'PREMIS_DESC', 'WEAPON_DESC',
       'STATUS_DESC', 'LOCATION', 'LAT', 'LON', 'VICTIM_RACE', 'SEVERITY'],
      dtype='object')

DR_NO  DATE_RPTD   DATE_OCC  TIME_OCC  AREA_NAME  CRM_1_SEVERITY  \
0    1307355 2010-02-20 2010-02-20  13:50:00     Newton               2   
1   11401303 2010-09-13 2010-09-12  00:45:00    Pacific               3   
2   70309629 2010-08-09 2010-08-09  15:15:00     Newton               1   
3   90631215 2010-01-05 2010-01-05  01:50:00  Hollywood               2   
4  100100501 2010-01-03 2010-01-02  21:00:00    Central               9   

   VICT_AGE VICT_SEX             PREMIS_DESC  \
0        48        M  SINGLE FAMILY DWELLING   
1         0        M                  STREET   
2         0        M                   ALLEY   
3        47        F                  STREET   
4        47        F                   ALLEY   

                                      WEAPON_DESC   STATUS_DESC  \
0                     UNKNOWN WEAPON/OTHER WEAPON  Adult Arrest   
1                     UNKNOWN WEAPON/OTHER WEAPON   Invest Cont   
2                     UNKNOWN WEAPON/OTHER WEAPON   Invest Cont   
3                                        HAND GUN   Invest Cont   
4  STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)   Invest Cont   

                                  LOCATION       LAT         LON  \
0   300 E  GAGE                         AV 33.982500 -118.269500   
1          SEPULVEDA                    BL 33.959900 -118.396200   
2  1300 E  21ST                         ST 34.022400 -118.252400   
3          CAHUENGA                     BL 34.101600 -118.329500   
4          8TH                          ST 34.038700 -118.248800   

              VICTIM_RACE  SEVERITY  
0  Hispanic/Latin/Mexican         2  
1                   White         3  
2  Hispanic/Latin/Mexican         1  
3                   White         3  
4  Hispanic/Latin/Mexican         9

In [39]:
display(df.info(null_counts = True))

## thus there are no null values

<ipython-input-39-bf8151f70309>:1: FutureWarning: null_counts is deprecated. Use show_counts instead
  display(df.info(null_counts = True))


<class 'pandas.core.frame.DataFrame'>
Int64Index: 2394173 entries, 0 to 276583
Data columns (total 16 columns):
 #   Column          Non-Null Count    Dtype         
---  ------          --------------    -----         
 0   DR_NO           2394173 non-null  int64         
 1   DATE_RPTD       2394173 non-null  datetime64[ns]
 2   DATE_OCC        2394173 non-null  datetime64[ns]
 3   TIME_OCC        2394173 non-null  object        
 4   AREA_NAME       2394173 non-null  category      
 5   CRM_1_SEVERITY  2394173 non-null  int32         
 6   VICT_AGE        2394173 non-null  int64         
 7   VICT_SEX        2394173 non-null  category      
 8   PREMIS_DESC     2394173 non-null  category      
 9   WEAPON_DESC     2394173 non-null  category      
 10  STATUS_DESC     2394173 non-null  category      
 11  LOCATION        2394173 non-null  category      
 12  LAT             2394173 non-null  float64       
 13  LON             2394173 non-null  float64       
 14  VICTIM_RACE     239

None

In [40]:
# Export to csv

df.to_csv("dataset.csv", index = False)

In [41]:
# export as pickle file

object_filename = "df.sav"

saved_object = pickle.dump(df, open(object_filename,'wb'))

print('Dataframe is saved into to disk successfully Using Pickle')

Dataframe is saved into to disk successfully Using Pickle


In [42]:
# to retrieve the file use:-

file = open("df.sav",'rb')
df = pickle.load(file)
file.close()